sdk/core/core-lro/src/poller.ts

```diff
@@ -317,7 +319,7 @@ export abstract class Poller<TState extends PollOperationState<TResult>, TResult
           abortSignal: options.abortSignal,
           fireProgress: this.fireProgress.bind(this)
         });
-        if (this.isDone() && this.resolve) {
+        if (this.isDone() && this.resolve && state.result) {
```

Comment:
This is a fundamental change. Why are we doing this? Please let me know if I'm missing something.

Reply:
The result has to be not `undefined` to be able to resolve with it. Do you think there are cases where we can resolve without a result?

Reply:
The root cause of the change is that `resolve` no longer accepts optional arguments.

Reply:
Well, yes! isDone only cares about `state.isCompleted || state.isCancelled || state.error`. If there's an error or if the poller is cancelled, result won't have a value. Can we use an exclamation mark instead? or something like that.

Reply:
This change breaks the communication phone number purchase LRO, see the timeout in CI.

This LRO has `void` as its `TResult` type, see https://github.com/Azure/azure-sdk-for-js/blob/351da5f682bca29a2516a6a1226c85c5b7a3bd02/sdk%2Fcommunication%2Fcommunication-administration%2Fsrc%2FphoneNumber%2Flro%2Fpurchase%2Fpoller.ts#L16

Instead of adding to the if condition here, can we change line 321 to `this.resolve(state.result!);` instead?

Reply:
Thanks @DominikMe for catching this! `state.result!` asserts that `state.result` can not be `undefined` which is not true if `TResult` is `void`. I cast `state.result` to `TResult` because we already know we finished polling so we either have a result or we do not if `TResult` is `void`. At the end of the day, both statements have the same behavior I believe but I think the latter approach makes the intent a bit more clear to the reader.

Also thanks for pointing out in our private communication that TS poller supports polling results to be of type `void` but .Net's does not. That was knew to me!